public/scripts/upload.js

```diff
@@ -46,41 +53,61 @@ async function runWorker(worker) {
 }
 
 function runPreviewWorker(worker, imageUpload) {
+  if (imageUpload.isErrorState) {
+    return;
+  }
+  window.viewModel.fileList.updateImageUploadStateByULID(
+    imageUpload.ulid,
+    ImageUploadState.GeneratingPreview,
+  );
   const imageUploadFile = imageUpload.file;
   worker.postMessage({
     imageUpload: imageUploadFile,
     location: window.location.toString(),
     deviceId: window.workerDeviceId,
     deviceInfo: window.deviceInfo,
+    ulid: imageUpload.ulid,
   });
   worker.addEventListener(
     "message",
     function (e) {
-      window.localforage
-        .setItem(`previewImage-${imageUpload.ulid}`, e.data[1])
-        .then(function () {
-          const imageContainer = document.querySelector(
-            ".upload__device-image-rect",
-          );
+      if (e.data["error"] !== undefined) {
+        console.log(
+          "Get error while generating preview image",
+          e.data["error"],
+        );
+        window.viewModel.fileList.updateImageUploadStateByULID(
+          e.data["ulid"],
+          ImageUploadState.ErrPreview,
+        );
+        return;
+      }
 
-          /* Put first generated mockup to preview area */
-          if (!imageContainer.style.backgroundImage) {
-            imageContainer.style.backgroundImage = `url(${e.data[1]})`;
-            imageContainer.style.backgroundSize = "cover";
-            imageContainer.style.backgroundPosition = "center";
-
-            const imageUploadHints = document.querySelectorAll(
-              ".upload__device-hint",
-            );
-            imageUploadHints.forEach((imageUploadHint) => {
-              imageUploadHint.innerHTML = "";
-              imageUploadHint.style.background = "transparent";
-            });
-          }
-        })
-        .catch(function (err) {
-          console.error("Get error while storing images to localforage:", err);
+      const imageContainer = document.querySelector(
+        ".upload__device-image-rect",
+      );
+
+      /* Put first generated mockup to preview area */
+      if (!imageContainer.style.backgroundImage) {
+        imageContainer.style.backgroundImage = `url(${e.data["results"][1]})`;
```

Comment:
```js
const someMeaningVar = e.data["results"]
// are we sure this array.length always >2? or need add guarding here?
const someOtherMeaningVar = someMeaningVar[1]
// or const [_, someOtherMeaningVar] = someMeaningVar
imageContainer.style.backgroundImage = `url(someOtherMeaningVar)`;

Reply:
@pkong-ds yes, the results should always be `tuple[int, int, int]`

Reply:
Yes, but the dev needs to beaware of that. That's why I prefer using some extra lines for sth like

```
const [_, previewUrl(?)] = e.data["results"]
```
```